README.md
# github.com/tiredofit/docker-osticket

[![GitHub release](https://img.shields.io/github/v/tag/tiredofit/docker-osticket?style=flat-square)](https://github.com/tiredofit/docker-osticket/releases/latest)
[![Build Status](https://img.shields.io/github/actions/workflow/status/tiredofit/docker-osticketmain.yml?branch=main&style=flat-square)](https://github.com/tiredofit/docker-osticket.git/actions)
[![Docker Stars](https://img.shields.io/docker/stars/tiredofit/osticket.svg?style=flat-square&logo=docker)](https://hub.docker.com/r/tiredofit/osticket/)
[![Docker Pulls](https://img.shields.io/docker/pulls/tiredofit/osticket.svg?style=flat-square&logo=docker)](https://hub.docker.com/r/tiredofit/osticket/)
[![Become a sponsor](https://img.shields.io/badge/sponsor-tiredofit-181717.svg?logo=github&style=flat-square)](https://github.com/sponsors/tiredofit)
[![Paypal Donate](https://img.shields.io/badge/donate-paypal-00457c.svg?logo=paypal&style=flat-square)](https://www.paypal.me/tiredofit)

* * *
## About

This will build a Docker Image for [OSTicket](https://www.osticket.com) - An open source helpdesk / ticketing system.

* Automatically installs and sets up installation upon first start

## Maintainer

- [Dave Conroy](https://github.com/tiredofit)

## Table of Contents


- [About](#about)
- [Maintainer](#maintainer)
- [Table of Contents](#table-of-contents)
- [Prerequisites and Assumptions](#prerequisites-and-assumptions)
- [Installation](#installation)
  - [Build from Source](#build-from-source)
  - [Prebuilt Images](#prebuilt-images)
- [Configuration](#configuration)
  - [Quick Start](#quick-start)
  - [Persistent Storage](#persistent-storage)
  - [Environment Variables](#environment-variables)
    - [Base Images used](#base-images-used)
  - [Networking](#networking)
- [Maintenance](#maintenance)
  - [Shell Access](#shell-access)
- [Support](#support)
  - [Usage](#usage)
  - [Bugfixes](#bugfixes)
  - [Feature Requests](#feature-requests)
  - [Updates](#updates)
- [License](#license)
- [References](#references)

## Prerequisites and Assumptions
*  Assumes you are using some sort of SSL terminating reverse proxy such as:
   *  [Traefik](https://github.com/tiredofit/docker-traefik)
   *  [Nginx](https://github.com/jc21/nginx-proxy-manager)
   *  [Caddy](https://github.com/caddyserver/caddy)
*  Requires access to a MySQL/MariaDB Server

## Installation

### Build from Source
Clone this repository and build the image with `docker build -t (imagename) .`

### Prebuilt Images
Builds of the image are available on [Docker Hub](https://hub.docker.com/r/tiredofit/osticket)

```bash
docker pull docker.io/tiredofit/osticket:(imagetag)
```

Builds of the image are also available on the [Github Container Registry](https://github.com/tiredofit/docker-osticket/pkgs/container/docker-osticket)

```
docker pull ghcr.io/tiredofit/docker-osticket:(imagetag)
```

The following image tags are available along with their tagged release based on what's written in the [Changelog](CHANGELOG.md):

| Container OS | Tag       |
| ------------ | --------- |
| Debian       | `:latest` |

## Configuration

### Quick Start

- The quickest way to get started is using [docker-compose](https://docs.docker.com/compose/). See the examples folder for a working [docker-compose.yml](examples/docker-compose.yml) that can be modified for development or production use.

- Set various [environment variables](#environment-variables) to understand the capabilities of this image.
- Map [persistent storage](#data-volumes) for access to configuration and data files for backup.
- Make [networking ports](#networking) available for public access if necessary

**The first boot can take from 2 minutes - 5 minutes depending on your CPU to setup the proper schemas.**

- Login to the web server and enter in your admin email address, admin password and start configuring the system!

### Persistent Storage
The following directories are used for configuration and can be mapped for persistent storage.

| Directory       | Description                                                                                 |
| --------------- | ------------------------------------------------------------------------------------------- |
| `/www/osticket` | (Not needed as we want to keep base clean, move to a custom/assets approach) Root Directory |
| `/www/logs`     | Nginx and php-fpm logfiles                                                                  |

### Environment Variables

#### Base Images used

This image relies on an [Alpine Linux](https://hub.docker.com/r/tiredofit/alpine) or [Debian Linux](https://hub.docker.com/r/tiredofit/debian) base image that relies on an [init system](https://github.com/just-containers/s6-overlay) for added capabilities. Outgoing SMTP capabilities are handlded via `msmtp`. Individual container performance monitoring is performed by [zabbix-agent](https://zabbix.com). Additional tools include: `bash`,`curl`,`less`,`logrotate`,`nano`.

Be sure to view the following repositories to understand all the customizable options:

| Image                                                         | Description                            |
| ------------------------------------------------------------- | -------------------------------------- |
| [OS Base](https://github.com/tiredofit/docker-debian/)        | Customized Image based on Debian Linux |
| [Nginx](https://github.com/tiredofit/docker-nginx/)           | Nginx webserver                        |
| [PHP-FPM](https://github.com/tiredofit/docker-nginx-php-fpm/) | PHP Interpreter                        |


| Parameter         | Description                                                                 | default                 |
| ----------------- | --------------------------------------------------------------------------- | ----------------------- |
| `INSTALL_SECRET`  | A Large and Random Installation String (Auto Generates on Install if empty) |                         |
| `INSTALL_EMAIL`   | Installer Email (Use different email then ADMIN_EMAIL)                      | `helpdesk@example.com`  |
| `INSTALL_NAME`    | Site Name                                                                   | `My Helpdesk`           |
| `ADMIN_FIRSTNAME` | First name of Admin User                                                    |                         |
| `ADMIN_LASTNAME`  | Last name of Admin User                                                     |                         |
| `ADMIN_EMAIL`     | Admin Email address (Make sure it is different than INSTALL_EMAIL)          |                         |
| `ADMIN_USER`      | Admin Username *Must be more than 5 characters*                             |                         |
| `ADMIN_PASS`      | Admin Password                                                              |                         |
| `CRON_INTERVAL`   | Amount of time in Minutes to Check Incoming Mail                            | `10`                    |
| `DB_HOST`         | Host or container name of MariaDB Server e.g. `osticket-db`                 |                         |
| `DB_PORT`         | MariaDB Port                                                                | `3306`                  |
| `DB_NAME`         | MariaDB Database name e.g. `osticket`                                       |                         |
| `DB_USER`         | MariaDB Username for above Database e.g. `osticket`                         |                         |
| `DB_PASS`         | MariaDB Password for above Database e.g. `password`                         |                         |
| `DB_PREFIX`       | Prefix for Tables                                                           | `ost_`                  |
| `SMTP_HOST`       | SMTP Host                                                                   | `postfix`               |
| `SMTP_PORT`       | SMTP Host Port                                                              | `25`                    |
| `SMTP_FROM`       | SMTP From Address                                                           | `osticket@hostname.com` |
| `SMTP_TLS`        | Should TLS be used (`0`=no `1`=yes)                                         | `1`                     |
| `SMTP_USER`       | SMTP Authentication user                                                    |                         |
| `SMTP_PASS`       | SMTP Authentication password                                                |                         |

### Networking

The following ports are exposed.

| Port | Description |
| ---- | ----------- |
| `80` | HTTP        |

* * *
## Maintenance

### Shell Access

For debugging and maintenance purposes you may want access the containers shell.

``bash
docker exec -it (whatever your container name is) bash
``
## Support

These images were built to serve a specific need in a production environment and gradually have had more functionality added based on requests from the community.
### Usage
- The [Discussions board](../../discussions) is a great place for working with the community on tips and tricks of using this image.
- Consider [sponsoring me](https://github.com/sponsors/tiredofit) for personalized support
### Bugfixes
- Please, submit a [Bug Report](issues/new) if something isn't working as expected. I'll do my best to issue a fix in short order.

### Feature Requests
- Feel free to submit a feature request, however there is no guarantee that it will be added, or at what timeline.
- Consider [sponsoring me](https://github.com/sponsors/tiredofit) regarding development of features.

### Updates
- Best effort to track upstream changes, More priority if I am actively using the image in a production environment.
- Consider [sponsoring me](https://github.com/sponsors/tiredofit) for up to date releases.

## License
MIT. See [LICENSE](LICENSE) for more details.

## References

* https://osticket.com



CHANGELOG.md
## 3.6.2 2023-07-30 <dave at tiredofit dot ca>

   ### Changed
      - Resolve issues with nginx configuration


## 3.6.1 2023-07-29 <jaydio@github>

   ### Changed
      - Fix spelling mistake in Dockerfile when enabling Zip extension


## 3.6.0 2023-07-28 <dave at tiredofit dot ca>

   ### Added
      - PHP 8.2 base

   ### Changed
      - Fix upstream issue with mysql command not being found due to changes in MariaDB client library
      - Rework writing cron scheduler


## 3.5.7 2023-07-27 <dave at tiredofit dot ca>

   ### Added
      - OS Ticket 1.18


## 3.5.6 2023-07-27 <dave at tiredofit dot ca>

   ### Added
      - Os Ticket 1.17.4


## 3.5.5 2023-06-05 <dekat@github>

   ### Changed
      - Fix for Admin name size check
      - Show errors on installation


## 3.5.4 2023-05-27 <dave at tiredofit dot ca>

   ### Changed
      - Fix for unterminated sed command


## 3.5.3 2023-04-14 <dave at tiredofit dot ca>

   ### Changed
      - Revert 3.5.2 and start doing 5 character checks on ADMIN_USER
      - Fix a problem with CRON_INTERVAL not being used


## 3.5.2 2023-04-14 <dave at tiredofit dot ca>

   ### Added
      - Add check for ADMIN_PASS size. Fail if <5 characters


## 3.5.1 2023-04-06 <dave at tiredofit dot ca>

   ### Added
      - Switch to PHP 8.1


## 3.5.0 2023-03-17 <dave at tiredofit dot ca>

   ### Added
      - Modernize Image
      - OsTicket 1.17.3


## 3.4.1 2022-06-23 <dave at tiredofit dot ca>

   ### Added
      - Support tiredofit/nginx:6.0.0 and tiredofit/nginx-php-fpm:7.0.0 changes


## 3.4.0 2021-11-23 <dave at tiredofit dot ca>

   ### Added
      - OSTIcket 1.15.4
      - Debian Bullseye base


## 3.3.3 2021-09-06 <eflukx@github>

   ### Fixed
      - Fix spelling mistake in Dockerfile for archiver plugin

## 3.3.2 2021-08-11 <cdhowie@github>

   ### Fixed
      - Clean NGINX_WEBROOT extra /www/ prefix from crontab

## 3.3.1 2021-08-11 <cdhowie@github>

   ### Changed
      - Cleanup composer cache
      - Cleanup Debian package cache to reduce image size 


## 3.3.0 2021-08-11 <dave at tiredofit dot ca>

   ### Added
      - Switch from Alpine to Debian base due to musl not supporting some functions OSTicket requires
      - Upgrade OSTicket to 1.15.3
      - PHP 7.4


## 3.2.1 2021-02-16 <leMail at github>

   ### Fixed
      - Removed erroneous argument in cron job


## 3.2.0 2020-06-15 <dave at tiredofit dot ca>

   ### Added
      - Update to support tiredofit/alpine 5.x.x base images


## 3.1.2 2020-01-02 <dave at tiredofit dot ca>

   ### Changed
      - Switch to php7-pecl-memcached


## 3.1.1 2020-01-02 <dave at tiredofit dot ca>

   ### Changed
      - Additional Changes to support new tiredofit/alpine base image


## 3.1.0 2019-12-29 <dave at tiredofit dot ca>

   ### Added
      - Support new tiredofit/nginx and tireofit/alpine base images


## 3.0.2 2019-12-17 <dave at tiredofit dot ca>

   ### Added
      - OSTicket 1.14.1
      - Refactored to support new tiredofit/nginx-php-fpm base image


## 3.0.1 2019-11-12 <dave at tiredofit dot ca>

* OSTicket 1.14-rc2

## 3.0 2019-09-12 <dave at tiredofit dot ca>

* Modernize Image
* Added many plugins
* OSTicket 1.14rc1
* PHP 7.3
* Alpine 3.10

## 2.8 2018-02-01 <dave at tiredofit dot ca>

* Pull sources from Github instead of mainwebsite
* Compile auth-ldap plugin

## 2.7 2018-02-01 <dave at tiredofit dot ca>

* Rebase

## 2.6 2017-10-05 <dave at tiredofit dot ca>

* Fix Broken Detection of new install

## 2.5 2017-08-29 <dave at tiredofit dot ca>

* Image Cleanup

## 2.4 2017-07-06 <dave at tiredofit dot ca>

* Added PHP_TIMEOUT

## 2.3 2017-07-03 <dave at tiredofit dot ca>

* Added PHP7-IMAP

## 2.2 2017-07-02 <dave at tiredofit dot ca>

* Build PHP7 Memcached Extension

## 2.1 2017-07-02 <dave at tiredofit dot ca>

* Sanity Checks in init scripts

## 2017-06-17 2.0 <dave at tiredofit dot ca>

* Rebase with nginx-php-fpm:7.0 with s6

## 2017-05-27 1.0 <dave at tiredofit dot ca>

* Production Stable
* Memcached Capable for Sessions
* Reset Admin Password if ENV Set upon Bootup


## 2017-05-27 0.9 <dave at tiredofit dot ca>

* Initial Release
* OSTicket 1.10
* Alpine:3.5
* PHP7
* Zabbix



install/assets/setup/install.php
<?php

//Configure settings from environmental variables
$_SERVER['HTTP_ACCEPT_LANGUAGE'] = getenv("LANGUAGE") ?: "en-us";

$vars = array(
  'name'          => getenv("INSTALL_NAME")  ?: 'My Helpdesk',
  'email'         => getenv("INSTALL_EMAIL") ?: 'helpdesk@example.com',
  'url'           => getenv("INSTALL_URL")   ?: 'http://localhost',
  'fname'         => getenv("ADMIN_FIRSTNAME"),
  'lname'         => getenv("ADMIN_LASTNAME"),
  'admin_email'   => getenv("ADMIN_EMAIL"),
  'username'      => getenv("ADMIN_USER"),
  'passwd'        => getenv("ADMIN_PASS"),
  'passwd2'       => getenv("ADMIN_PASS"),
  'prefix'        => getenv("DB_PREFIX")               ?: 'ost_',
  'dbhost'        => getenv("DB_HOST"),             
  'dbname'        => getenv("DB_NAME"),            
  'dbuser'        => getenv("DB_USER"),
  'dbport'        => getenv("DB_PORT")              ?: '3306',
  'dbpass'        => getenv("DB_PASS")              ?: getenv("DB_PASS"),
  'smtp_host'     => getenv("SMTP_HOST")            ?: 'postfix-relay',
  'smtp_port'     => getenv("SMTP_PORT")            ?: 25,
  'smtp_from'     => getenv("SMTP_FROM"),
  'smtp_tls'      => getenv("SMTP_TLS"),
  'smtp_tls_certs' => getenv("SMTP_TLS_CERTS")       ?: '/etc/ssl/certs/ca-certificates.crt',
  'smtp_user'     => getenv("SMTP_USER"),
  'smtp_pass'     => getenv("SMTP_PASSWORD"),
  'cron_interval'  => getenv("CRON_INTERVAL")        ?: 5,
  'siri'          => getenv("INSTALL_SECRET"),
  'config'        => getenv("INSTALL_CONFIG") ?: '<WEBROOT>/include/ost-sampleconfig.php'
);

//Script settings
define('CONNECTION_TIMEOUT_SEC', 180);
function err( $msg) {
  fwrite(STDERR, "$msg\n");
  exit(1);
}
function boolToOnOff($v) {
  return ((boolean) $v) ? 'on' : 'off';
}
function convertStrToBool($varName, $default) {
  global $vars;
   if ($vars[$varName] != '') {
     return $vars[$varName] == '1';
   }
   return $default;
}

// Override Helpdesk URL. Only applied during database installation.
define("URL",$vars['url']);

//Require files (must be done before any output to avoid session start warnings)
chdir("<WEBROOT>/setup_hidden");
require "<WEBROOT>/setup_hidden/setup.inc.php";
require_once INC_DIR.'class.installer.php';


/************************* Mail Configuration *******************************************/
define('MAIL_CONFIG_FILE','/etc/msmtp');

echo "** [osticket] Configuring mail settings\n";
if (!$mailConfig = file_get_contents('/assets/msmtp/msmtp.conf')) {
  err("** [osticket] Failed to load mail configuration file");
};
$mailConfig = str_replace('%SMTP_HOSTNAME%', $vars['smtp_host'], $mailConfig);
$mailConfig = str_replace('%SMTP_PORT%', $vars['smtp_port'], $mailConfig);
$v = !empty($vars['smtp_from']) ? $vars['smtp_from'] : $vars['smtp_user'];
$mailConfig = str_replace('%SMTP_FROM%', $v, $mailConfig);
$mailConfig = str_replace('%SMTP_USER%', $vars['smtp_user'], $mailConfig);
$mailConfig = str_replace('%SMTP_PASS%', $vars['smtp_pass'], $mailConfig);
$mailConfig = str_replace('%SMTP_TLS_CERTS%', $vars['smtp_tls_certs'], $mailConfig);

$mailConfig = str_replace('%SMTP_TLS%', boolToOnOff(convertStrToBool('smtp_tls',true)), 
$mailConfig);
$mailConfig = str_replace('%SMTP_AUTH%', boolToOnOff($vars['smtp_user'] != ''), $mailConfig);

if (!file_put_contents(MAIL_CONFIG_FILE, $mailConfig) || !chown(MAIL_CONFIG_FILE,'nginx')
   || !chgrp(MAIL_CONFIG_FILE,'www-data') || !chmod(MAIL_CONFIG_FILE,0600)) {
   err("Failed to write mail configuration file");
}


/************************* OSTicket Installation *******************************************/

//Create installer class
define('OSTICKET_CONFIGFILE','<WEBROOT>/include/ost-config.php');
$installer = new Installer(OSTICKET_CONFIGFILE); //Installer instance.


// Always set mysqli.default_port for osTicket db_connect
ini_set('mysqli.default_port', $vars['dbport']);

//Check database installation status
$db_installed = false;
echo "** [osticket] DB - Connecting to database mysql://${vars['dbuser']}@${vars['dbhost']}/${vars['dbname']}\n";
if (!db_connect($vars['dbhost'],$vars['dbuser'],$vars['dbpass']))
   err(sprintf(__('** [osticket] Unable to connect to MySQL server: %s'), db_connect_error()));
elseif(explode('.', db_version()) < explode('.', $installer->getMySQLVersion()))
   err(sprintf(__('** [osticket] osTicket requires MySQL %s or later!'),$installer->getMySQLVersion()));
elseif(!db_select_database($vars['dbname']) && !db_create_database($vars['dbname'])) {
   err("** [osticket] Database doesn't exist");
} elseif(!db_select_database($vars['dbname'])) {
   err('** [osticket] Unable to select the database');
} else {
   $sql = 'SELECT * FROM `'.$vars['prefix'].'config` LIMIT 1';
   if(db_query($sql, false)) {
       $db_installed = true;
       echo "** [osticket] Database already installed\n";
   }
}

//Create secret if not set by env var and not previously stored
DEFINE('SECRET_FILE','<WEBROOT>/secret.txt');
if (!$vars['siri']) {
  if (file_exists(SECRET_FILE)) {
    echo "** [osticket] DB - Loading installation secret\n";
    $vars['siri'] = file_get_contents(SECRET_FILE);
  } else {
    echo "** [osticket] DB - Generating new installation secret and saving\n";
    //Note that this randomly generated value is not intended to secure production sites!
    $vars['siri'] = 
substr(str_shuffle("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890_="), 0, 64);
    file_put_contents(SECRET_FILE, $vars['siri']);
  }
} else {
  echo "** [osticket] DB - Using installation secret from INSTALL_SECRET environmental variable\n";
}

//Always rewrite config file in case MySQL details changed (e.g. ip address)
echo "** [osticket] DB - Updating configuration file\n";
if (!$configFile = file_get_contents($vars['config'])) {
  err("** [osticket] DB - Failed to load configuration file: {$vars['config']}");
};
$configFile= 
str_replace("define('OSTINSTALLED',FALSE);","define('OSTINSTALLED',TRUE);",$configFile);
$configFile= str_replace('%ADMIN-EMAIL',$vars['admin_email'],$configFile);
$configFile= str_replace('%CONFIG-DBHOST',$vars['dbhost'],$configFile);
$configFile= str_replace('%CONFIG-DBNAME',$vars['dbname'],$configFile);
$configFile= str_replace('%CONFIG-DBUSER',$vars['dbuser'],$configFile);
$configFile= str_replace('%CONFIG-DBPASS',$vars['dbpass'],$configFile);
$configFile= str_replace('%CONFIG-PREFIX',$vars['prefix'],$configFile);
$configFile= str_replace('%CONFIG-SIRI',$vars['siri'],$configFile);

if (!file_put_contents($installer->getConfigFile(), $configFile)) {
   err("** [osticket] DB - Failed to write configuration file");
}

//Perform database installation if required
if (!$db_installed) {
  echo "** [osticket] DB - Installing database. Please wait...\n";
  if (!$installer->install($vars)) {
    $errors=$installer->getErrors();
    echo "** [osticket] DB - Database installation failed. Errors:\n";
    foreach($errors as $e) {
      echo "  $e\n";
    }
    exit(1);
  } else {
    echo "** [osticket] DB - Database installation successful\n";
  }
}

?>
